Building a react-native Android project fails

Is anyone using Nixos to build react-native projects?

The command `./gradlew build --stacktrace` fails with error

```
com.android.builder.internal.aapt.v2.Aapt2InternalException: AAPT2 aapt2-3.4.1-5326820-linux 
Daemon #0: Daemon startup failed
```
And a some lines below:
```
Caused by: java.io.IOException: Cannot run program "/home/federico/.gradle/caches/transforms-2/files-2.1/2df53a66e454f1abff0680fcd8c73763/aapt2-3.4.1-5326820-linux/aapt2": error=2, No such file or directory
```
However said file is there:
```
 $ file /home/federico/.gradle/caches/transforms-2/files-2.1/2df53a66e454f1abff0680fcd8c73763/aapt2-3
.4.1-5326820-linux/aapt2
/home/federico/.gradle/caches/transforms-2/files-2.1/2df53a66e454f1abff0680fcd8c73763/aapt2-3.4.1-5326820-linux/aapt2: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.24, not stripped
```
I suspect this is related to linking, any hints?

--- a/doc/release-notes/rl-2605.section.md
+++ b/doc/release-notes/rl-2605.section.md
@@ -26,6 +26,8 @@
   If your SQLite database is corrupted, the migration might fail and require [manual intervention](https://github.com/louislam/uptime-kuma/issues/5281).
   See the [migration guide](https://github.com/louislam/uptime-kuma/wiki/Migration-From-v1-To-v2) for more information.
 
+- We now use the upstream wrapper script for Gradle, supporting both the `JAVA_HOME` and `GRADLE_OPTS` environment variables.
+
 ## Nixpkgs Library {#sec-nixpkgs-release-26.05-lib}
 
 <!-- To avoid merge conflicts, consider adding your item at an arbitrary place in the list instead. -->

--- a/pkgs/development/tools/build-managers/gradle/default.nix
+++ b/pkgs/development/tools/build-managers/gradle/default.nix
@@ -68,8 +68,8 @@ let
               ''
                 cp -a $src/* .
                 substituteInPlace ./build.gradle --replace-fail '@JAVA_VERSION@' '${javaMajorVersion}'
-                env GRADLE_USER_HOME=$TMPDIR/gradle org.gradle.native.dir=$TMPDIR/native \
-                gradle run --no-daemon --quiet --console plain > $out
+                env GRADLE_USER_HOME=$TMPDIR/gradle GRADLE_OPTS=-Dorg.gradle.native.dir=$TMPDIR/native \
+                  gradle run --no-daemon --quiet --console plain > $out
                 actual="$(<$out)"
                 if [[ "${javaVersion}" != "$actual"* ]]; then
                   echo "Error: Expected '${javaVersion}', to start with '$actual'" >&2
@@ -275,7 +275,7 @@ let
         version = testers.testVersion {
           package = finalAttrs.finalPackage;
           command = ''
-            env GRADLE_USER_HOME=$TMPDIR/gradle org.gradle.native.dir=$TMPDIR/native \
+            env GRADLE_USER_HOME=$TMPDIR/gradle GRADLE_OPTS=-Dorg.gradle.native.dir=$TMPDIR/native \
               gradle --version
           '';
         };
@@ -291,7 +291,11 @@ let
               }
               ''
                 cp -a $src/* .
-                env GRADLE_USER_HOME=$TMPDIR/gradle org.gradle.native.dir=$TMPDIR/native \
+
+                # Make sure GRADLE_OPTS works.
+                env \
+                  GRADLE_USER_HOME=$TMPDIR/gradle \
+                  GRADLE_OPTS="-Dorg.gradle.native.dir=$TMPDIR/native -Dnix.test.mainClass=Main" \
                   gradle run --no-daemon --quiet --console plain > $out
               '';
         };

--- a/pkgs/development/tools/build-managers/gradle/default.nix
+++ b/pkgs/development/tools/build-managers/gradle/default.nix
@@ -129,8 +129,11 @@ let
       callPackage,
       makeWrapper,
       unzip,
+      coreutils,
+      findutils,
       ncurses5,
       ncurses6,
+      gnused,
       udev,
       testers,
       runCommand,
@@ -176,28 +179,47 @@ let
       # We only need to patchelf some libs embedded in JARs.
       dontAutoPatchelf = true;
 
+      # All the installed Gradle libraries and binaries go here. Note that the Gradle wrapper
+      # will look for a lib directory above its own directory (which is presumed to be bin)
+      # for loading the main Gradle jar.
+      gradleLibexec = "${placeholder "out"}/libexec/gradle";
+
       installPhase =
-        with builtins;
         let
           # set toolchains via installations.path property in gradle.properties.
           # See https://docs.gradle.org/current/userguide/toolchains.html#sec:custom_loc
-          toolchainPaths = "org.gradle.java.installations.paths=${concatStringsSep "," javaToolchains}";
-          jnaLibraryPath = if stdenv.hostPlatform.isLinux then lib.makeLibraryPath [ udev ] else "";
-          jnaFlag =
-            if stdenv.hostPlatform.isLinux then "--add-flags \"-Djna.library.path=${jnaLibraryPath}\"" else "";
+          toolchainPaths = "org.gradle.java.installations.paths=${lib.concatStringsSep "," javaToolchains}";
+          jnaLibraryPath = lib.optionalString stdenv.hostPlatform.isLinux (lib.makeLibraryPath [ udev ]);
+          jnaFlag = lib.optionalString stdenv.hostPlatform.isLinux ''--add-flags "-Djna.library.path=${jnaLibraryPath}"'';
         in
         ''
-          mkdir -pv $out/lib/gradle/
-          cp -rv lib/ $out/lib/gradle/
+          # Install all the Gradle jars.
+          mkdir -vp $gradleLibexec
+          cp -av lib/ $gradleLibexec
+          [ -f $gradleLibexec/lib/gradle-launcher-*.jar ] || { echo "No Gradle launcher jar found!" >&2 && exit 1; }
 
-          gradle_launcher_jar=$(echo $out/lib/gradle/lib/gradle-launcher-*.jar)
-          test -f $gradle_launcher_jar
-          makeWrapper ${java}/bin/java $out/bin/gradle \
-            --set JAVA_HOME ${java} \
-            ${jnaFlag} \
-            --add-flags "-classpath $gradle_launcher_jar org.gradle.launcher.GradleMain"
+          # Set up toolchain paths in the gradle.properties.
+          echo ${lib.escapeShellArg toolchainPaths} > $gradleLibexec/gradle.properties
 
-          echo "${toolchainPaths}" > $out/lib/gradle/gradle.properties
+          # Just use the existing gradle wrapper (usually called `gradlew`) as the main program.
+          mkdir -vp $gradleLibexec/bin
+          cp -v bin/gradle $gradleLibexec/bin/gradlew
+          chmod +x $gradleLibexec/bin/gradlew
+          patchShebangs --host $gradleLibexec/bin/gradlew
+
+          # Ensure that JAVA_HOME is set so the installed Gradle wrapper picks it up.
+          # The wrapper also needs coreutils, xargs, and sed.
+          mkdir -vp $out/bin
+          makeWrapper $gradleLibexec/bin/gradlew $out/bin/gradle \
+            --set-default JAVA_HOME ${java} \
+            --suffix PATH : ${
+              lib.makeBinPath [
+                coreutils
+                findutils
+                gnused
+              ]
+            } \
+            ${jnaFlag}
         '';
 
       dontFixup = !stdenv.hostPlatform.isLinux;
@@ -212,10 +234,10 @@ let
           export PATH="${buildPackages.jdk}/bin:$PATH"
           . ${./patching.sh}
 
-          nativeVersion="$(extractVersion native-platform $out/lib/gradle/lib/native-platform-*.jar)"
+          nativeVersion="$(extractVersion native-platform $gradleLibexec/lib/native-platform-*.jar)"
           for variant in "" "-ncurses5" "-ncurses6"; do
             autoPatchelfInJar \
-              $out/lib/gradle/lib/native-platform-linux-${arch}$variant-''${nativeVersion}.jar \
+              $gradleLibexec/lib/native-platform-linux-${arch}$variant-''${nativeVersion}.jar \
               "${lib.getLib stdenv.cc.cc}/lib64:${
                 lib.makeLibraryPath [
                   stdenv.cc.cc
@@ -228,14 +250,14 @@ let
           # The file-events library _seems_ to follow the native-platform version, but
           # we won’t assume that.
           if [ -n "${newFileEvents}" ]; then
-            fileEventsVersion="$(extractVersion gradle-fileevents $out/lib/gradle/lib/gradle-fileevents-*.jar)"
+            fileEventsVersion="$(extractVersion gradle-fileevents $gradleLibexec/lib/gradle-fileevents-*.jar)"
             autoPatchelfInJar \
-              $out/lib/gradle/lib/gradle-fileevents-''${fileEventsVersion}.jar \
+              $gradleLibexec/lib/gradle-fileevents-''${fileEventsVersion}.jar \
               "${lib.getLib stdenv.cc.cc}/lib64:${lib.makeLibraryPath [ stdenv.cc.cc ]}"
           else
-            fileEventsVersion="$(extractVersion file-events $out/lib/gradle/lib/file-events-*.jar)"
+            fileEventsVersion="$(extractVersion file-events $gradleLibexec/lib/file-events-*.jar)"
             autoPatchelfInJar \
-              $out/lib/gradle/lib/file-events-linux-${arch}-''${fileEventsVersion}.jar \
+              $gradleLibexec/lib/file-events-linux-${arch}-''${fileEventsVersion}.jar \
               "${lib.getLib stdenv.cc.cc}/lib64:${lib.makeLibraryPath [ stdenv.cc.cc ]}"
           fi
 

--- a/pkgs/development/tools/build-managers/gradle/tests/java-application/build.gradle
+++ b/pkgs/development/tools/build-managers/gradle/tests/java-application/build.gradle
@@ -2,6 +2,14 @@ plugins {
     id('application')
 }
 
+String getPropOrDie(String prop, String message) {
+    def ret = System.getProperty(prop)
+    if (!ret) {
+        throw new RuntimeException(message)
+    }
+    return ret
+}
+
 application {
-    mainClass = 'Main'
+    mainClass = getPropOrDie('nix.test.mainClass', 'No main class defined, perhaps GRADLE_OPTS broke?')
 }

